### PR TITLE
Swift 3.0 with 2.x Compatability

### PR DESCRIPTION
**Pull request by ahtierney**
Migrated to 3.0 and added backward compatibility to 2.x where applicable.

TODOs:
- [x] Migrate to swift 3
- [x] Support for swift 2.3 
- [x] Support for 2.x and iOS 9

### CODE REVIEW

File: BoardingPass.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -386,7 +395,7 @@
 			);
 			runOnlyForDeploymentPostprocessing = 0;
 			shellPath = /bin/sh;
-			shellScript = "if which swiftlint >/dev/null; then\nswiftlint\nelse\necho \"warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint\"\nfi";
+			shellScript = "exit 0\nif which swiftlint >/dev/null; then\nswiftlint\nelse\necho \"warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint\"\nfi";
```
Comment (ateliercw) [inline BoardingPass.xcodeproj/project.pbxproj]: #req re-enable swiftlint, fix the swiftlint warnings / errors

File: BoardingPass/BoardingInformation.swift [hunk 2]
```diff
@@ -80,11 +86,14 @@ public extension BoardingInformation {
...
-        guard let index = viewControllers.indexOf(viewController) where index > 0 else {
+    func viewController(beforeController viewController: UIViewController) -> UIViewController? {
+        guard let index = viewControllers.index(of: viewController) else {
+            return nil
+        }
+        guard index > 0 else {
             return nil
         }
-        return viewControllers[index.predecessor()]
+        return viewControllers[(index - 1)]
```
Comment (ateliercw) [inline BoardingPass/BoardingInformation.swift]: #pp `return viewControllers[index.advanced(by: -1)] `
totally a personal style preference, I've been using those for index manipulation

File: BoardingPass/BoardingNavigationController.swift [hunk 3]
```diff
@@ -91,43 +161,43 @@ public class BoardingNavigationController: UINavigationController {
      - parameter animated: Specify true to animate the transition or false
      if you do not want the transition to be animated.
      */
-    public func popToPreviousViewController(animated animated: Bool) {
-        if let poppableViewController = topViewController.flatMap(boardingInfoAfter) {
-            popToAndInsertIfNeeded(poppableViewController, animated: animated)
+    public func popToPreviousViewController(animated: Bool) {
+        if let poppableViewController = topViewController.flatMap({ boardingInfo(afterController: $0) }) {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #? I think `.flatMap( boardingInfo(afterController: ))` will work without having to wrap it in a closure

File: BoardingPass/BoardingNavigationController.swift [hunk 4]
```diff
@@ -79,9 +149,9 @@ public class BoardingNavigationController: UINavigationController {
      - parameter animated: Specify true to animate the transition or false
      if you do not want the transition to be animated.
      */
-    public func pushToNextViewController(animated animated: Bool) {
-        if let pushableViewController = topViewController.flatMap(boardingInfoAfter) {
-            pushViewController(pushableViewController, animated: animated)
+    public func pushToNextViewController(animated isAnimated: Bool) {
+        if let pushableViewController = topViewController.flatMap({ boardingInfo(afterController: $0) }) {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #? I think `.flatMap( boardingInfo(afterController: ))` will work without having to wrap it in a closure

File: BoardingPass/BoardingNavigationController.swift [hunk 5]
```diff
@@ -137,102 +207,114 @@ private extension BoardingNavigationController {
...
-        return viewControllersToPresent[index.predecessor()]
+        guard index > 0 else {
+            return nil
+        }
+
+        return viewControllersToPresent[(index - 1)]
     }
 
-    func viewControllerAfter(viewController: UIViewController?) -> UIViewController? {
+    func viewController(afterController viewController: UIViewController?) -> UIViewController? {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #pp rename to`viewController(after viewController: UIViewController?)`

File: BoardingPass/BoardingNavigationController.swift [hunk 6]
```diff
@@ -137,102 +207,114 @@ private extension BoardingNavigationController {
             finishAnimation(forRecognizer: sender)
         }
     }
+#endif
 }
 
 private extension BoardingNavigationController {
 
-    func viewControllerBefore(viewController: UIViewController?) -> UIViewController? {
+    func viewController(beforeController viewController: UIViewController?) -> UIViewController? {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #pp rename to `viewController(before viewController: UIViewController?)`

File: BoardingPass/BoardingNavigationController.swift [hunk 7]
```diff
@@ -91,43 +161,43 @@ public class BoardingNavigationController: UINavigationController {
...
-    public func navigationController(navigationController: UINavigationController, interactionControllerForAnimationController animationController: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
-        return interactionController
-    }
-}
-
-// MARK: - Actions
+// MARK: Actions
 private extension BoardingNavigationController {
-    func popToAndInsertIfNeeded(viewController: UIViewController, animated: Bool) {
+    func popToAndInsertIfNeeded(controller viewController: UIViewController, animated: Bool) {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #pp rename to `popToAndInsertIfNeeded(_ viewController: UIViewController, animated: Bool)`

File: BoardingPass/BoardingNavigationController.swift [hunk 8]
```diff
@@ -137,102 +207,114 @@ private extension BoardingNavigationController {
...
         }
         if interactionController == nil {
             interactionController = UIPercentDrivenInteractiveTransition()
         }
         // The transitioning delegate being nil tells us that there isn't another active transition in play
-        if transitionState.direction == .None && transitioningDelegate == nil {
+        if transitionState.direction == .none && transitioningDelegate == nil {
             if xTranslation < 0 {
-                guard let pushableViewControler = topViewController.flatMap(boardingInfoAfter) else {
+                guard let pushableViewControler = topViewController.flatMap({ boardingInfo(afterController: $0) }) else {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #? should be able to use the function without wrapping in a closure here

File: BoardingPass/BoardingNavigationController.swift [hunk 9]
```diff
@@ -137,102 +207,114 @@ private extension BoardingNavigationController {
...
                     return
                 }
-                transitionState = TransitionState(direction: .Push, previousState: viewControllers)
+                transitionState = TransitionState(direction: .push, previousState: viewControllers)
                 pushViewController(pushableViewControler, animated: true)
             }
             else if xTranslation > 0 {
-                guard let poppableViewController = topViewController.flatMap(boardingInfoBefore) else {
+                guard let poppableViewController = topViewController.flatMap({ boardingInfo(beforeController: $0) }) else {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #? should be able to use the function without wrapping in a closure

File: BoardingPassExample/Onboarding Demo/DemoCompatibility.swift [hunk 10]
```diff
@@ -0,0 +1,101 @@
+//
```
Comment (ateliercw) [inline BoardingPassExample/Onboarding Demo/DemoCompatibility.swift]: #? do we need Swift 2.3 cross compatibility in the demo, I'd much rather see a Swift 2.3 demo and a Swift 3 demo

**Comment by dostrander** [comment]
@ateliercw Just updated it with the feedback you did a while ago. Trying to push this through so we can start using it next week if you can help out :)

File: BoardingPass/BoardingNavigationController.swift [hunk 11]
```diff
@@ -91,43 +161,43 @@ public class BoardingNavigationController: UINavigationController {
      - parameter animated: Specify true to animate the transition or false
      if you do not want the transition to be animated.
      */
-    public func popToPreviousViewController(animated animated: Bool) {
-        if let poppableViewController = topViewController.flatMap(boardingInfoAfter) {
+    public func popToPreviousViewController(animated: Bool) {
```
Comment (ateliercw) [inline BoardingPass/BoardingNavigationController.swift]: #req add isAnimated to make this line up with the `pishToNextViewController` function

**Pull request merged**